lib/kubernetes-deploy/discoverable_resource.rb

```diff
@@ -0,0 +1,175 @@
+    STATUS_SUCCESS_ANNOTATION = 'kubernetes-deploy.shopify.io/status-success'
+    TIMEOUT_ANNOTATION = 'kubernetes-deploy.shopify.io/timeout'
+    PREDEPLOY_ANNOTATION = 'kubernetes-deploy.shopify.io/predeploy'
+    PRUNABLE_ANNOTATION = 'kubernetes-deploy.shopify.io/prunable'
+
+    def type
+      self.class.type
+    end
+
+    def self.inherited(child_class)
+      child_classes.add child_class
+    end
+
+    def self.child_classes
+      @child_classes ||= Set.new
+    end
+
+    def self.prunable?
+      return @prunable if defined?(@prunable)
+      false
+    end
+
+    def self.predeploy?
+      return @predeploy if defined?(@predeploy)
+      false
+    end
+
+    def self.timeout
+      return @timeout if defined?(@timeout)
+      super
+    end
+
+    def self.identity
+      "#{group}/#{version}/#{type}"
+    end
+
+    def self.all
+      child_classes.dup
+    end
+
+    def self.discover(context:, logger:)
+      logger.info("Discovering custom resources:")
+      @resources = nil
+      @child_classes = nil
+      discover_tpr(v1beta1_kubeclient(context))
+      begin
+        discover_crd(v1beta1_crd_kubeclient(context))
+      rescue KubeException => err
+        logger.warn("Unable to discover CustomResourceDefinitions: #{err}")
+      end
+    end
+
+    def self.build(namespace:, context:, definition:, logger:)
+      # We only discover once per kubernetes-deploy invocation
+      discover(context: context, logger: logger) unless @resources
+
+      type = definition["kind"]
+      group, _, version = definition['apiVersion'].rpartition('/')
+
+      resource_class = @resources.dig(group, type, version) if @resources
+      return super unless resource_class
+
+      opts = { namespace: namespace, context: context, definition: definition, logger: logger }
+      resource_class.new(**opts)
+    end
+
+    def self.discover_tpr(client)
+      return unless client.respond_to? :get_third_party_resources
+      resources = client.get_third_party_resources
+      resources.each do |res|
+        type, _, group = res.metadata.name.partition('.')
+        # TPR API supports multiple versions for a single resource :(
+        res.versions.each do |version|
+          discovered(group: group,
+                     type: type,
+                     version: version.name,
+                     annotations: res.metadata.annotations)
+        end
+      end
+    end
+
+    def self.discover_crd(client)
+      return unless client.respond_to? :get_custom_resource_definitions
+      resources = client.get_custom_resource_definitions
+      resources.each do |res|
+        discovered(group: res.spec.group,
+                   type: res.spec.names.kind,
+                   version: res.spec.version,
+                   annotations: res.metadata.annotations)
+      end
+    end
+
+    def self.discovered(group:, type:, version:, annotations:)
+      resource_class = Class.new(self) do
```

Comment:
I previously spoke about using meta-programming with @mkobetic. Since that time I've convinced myself this is the right approach.

We are modelling a class/instance relationship (resource definition vs resource instances). Each definition has specific properties shared by all instances (e.g. whether it's prunable, or its configured timeout). This information should be stored somewhere, since Ruby supports OOP it makes sense to generate classes that can contain the specific settings shared by all resources (instances).

Reply:
**DISCLAIMER:** I've only worked with this kind of thing a couple times in the past, and I'm going to want to pull in another experienced Rubyist before we merge this PR.

Agreed, but I think we can take it a step further and make the generated classes "normal" rather than using class variables, by using `class_eval`. Here's a working example:

```ruby 
module ScratchPadEphemeral
  class DynamicParent
    require 'active_support/core_ext/class/subclasses'

    def self.define_child(child_name, child_type, some_bool)
      new_class = self.class_eval <<-STRING
        class #{child_name.capitalize} < DynamicParent
          def type
            '#{child_type}'
          end

          def some_bool
            #{some_bool}
          end
        end
      STRING
    end
  end
end
```

```
[35] pry(ScratchPadEphemeral):1> DynamicParent.define_child("foo", "test", true)
=> :some_bool
[36] pry(ScratchPadEphemeral):1> DynamicParent.define_child("bar", "test", false)
=> :some_bool
[37] pry(ScratchPadEphemeral):1> DynamicParent.subclasses
=> [ScratchPadEphemeral::DynamicParent::Bar, ScratchPadEphemeral::DynamicParent::Foo]
[38] pry(ScratchPadEphemeral):1> foo = DynamicParent::Foo.new
=> #<ScratchPadEphemeral::DynamicParent::Foo:0x007fd8e4106fe8>
[39] pry(ScratchPadEphemeral):1> foo.type
=> "test"
[40] pry(ScratchPadEphemeral):1> foo.some_bool
=> true
```

This should incidentally be more performant than `define_method?`, from what I've been told in the past.

Reply:
Adopted this approach in the rewritten version, thanks.